sycl/doc/syclcompat/README.md

```diff
@@ -1729,6 +1729,140 @@ static void invoke_kernel_function(kernel_function &function,
 
 } // namespace syclcompat
 ```
+The kernel_launcher class provides a utility for launching SYCL kernels
+through function wrappers, managing execution configurations such as SYCL
+queues, nd_range, and local memory size as thread-local variables. It allows
+kernel invocation through multiple overloads of launch, either by directly
+calling a function pointer with arguments, invoking a registered function
+wrapper, or using packed argument arrays for flexible invocation. The class
+also maintains a mapping of raw function pointers to launchable functors,
+enabling dynamic function registration.
+
+The wrapper_register class serves as a helper for registering kernel functions
+and invoking them through a wrapper. It supports constructing from a kernel
+function pointer,overloading operator() for kernel execution, and retrieving
+the original function pointer via get(), providing a mechanism for managing
+and invoking SYCL kernels dynamically.
+
+``` c++
+namespace syclcompat {
+/// Utility class for launching SYCL kernels through kernel
+/// function wrapper.
+/// For example:
+/// A SYCL kernel function:
+///   void kernel_func(int *ptr, sycl::nd_item<3> item);
+/// Kernel function wrapper:
+///   void kernel_func_wrapper(int *ptr) {
+///     sycl::queue queue = *syclcompat::kernel_launcher::_que;
+///     unsigned int localMemSize =
+///     syclcompat::kernel_launcher::_local_mem_size;
+///     sycl::nd_range<3> nr =
+///     syclcompat::kernel_launcher::_nr;
+///
+///     queue.parallel_for(
+///       nr,
+///       [=](sycl::nd_item<3> item_ct1) {
+///         kernel_func(ptr, item_ct1);
+///       });
+///   }
+/// Then launch the kernel through wrapper like:
+///   typedef void(*fpt)(int *);
+///   fpt fp = kernel_func_wrapper;
+///   syclcompat::kernel_launcher::launch(fp, syclcompat::dim3(1),
+///     syclcompat::dim3(1), 0, 0, device_ptr);
+/// If the origin function type is erased, then need to register it first:
+///   void *fp = (void *)wrapper_register(&kernel_func_wrapper).get();
+///   syclcompat::kernel_launcher::launch(fp, syclcompat::dim3(1),
+///     syclcompat::dim3(1), args, 0, 0);
+class kernel_launcher {
+public:
+  /// Variables for storing execution configuration.
+  static inline thread_local sycl::queue *_que = nullptr;
+  static inline thread_local sycl::nd_range<3> _nr = sycl::nd_range<3>();
+  static inline thread_local unsigned int _local_mem_size = 0;
+  /// Map for retrieving launchable functor from a raw pointer.
+  static inline std::map<
+      const void *,
+      std::function<void(dim3, dim3, void **, unsigned int, queue_ptr)>>
+      kernel_function_ptr_map;
```

Comment:
I have the following concerns about these:

1. The naming does not seem to correspond to other public members in syclcompat. I.e. starting with an underscore is not commonplace from what I can tell, except for maybe `kernel_function_ptr_map`.
2. Exposing these members directly allows for user-side tampering. Based on the examples it seems to me like they are just for getting and users should not be able to set them. Can we maybe make these private and add getters for them?
3. Does the user actually need direct access to `kernel_function_ptr_map`? In the implementation it seems to be guarded by a private mutex, so I would assume any direct access to it from user-side would be invalid.